### PR TITLE
Generate Parameter Library for Joint Trajectory Controller

### PR DESCRIPTION
**Pull request by tylerjw**
Signed-off-by: Tyler Weaver <maybe@tylerjw.dev>

This is an PR to converting the parameter usage in JTC to use https://github.com/PickNikRobotics/generate_parameter_library

### CODE REVIEW

File: joint_trajectory_controller/package.xml [hunk 1]
```diff
@@ -14,6 +14,7 @@
   <build_depend>angles</build_depend>
   <build_depend>pluginlib</build_depend>
   <build_depend>realtime_tools</build_depend>
+  <build_depend>generate_parameter_library</build_depend>
```
Comment (AndyZe) [inline joint_trajectory_controller/package.xml:17]: Alphabetize

File: joint_trajectory_controller/include/joint_trajectory_controller/validate_joint_trajectory_controller_parameters.hpp [hunk 2]
```diff
@@ -0,0 +1,55 @@
+Result command_interface_type_combinations(rclcpp::Parameter const& parameter) {
```
Comment (AndyZe) [inline joint_trajectory_controller/include/joint_trajectory_controller/validate_joint_trajectory_controller_parameters.hpp:1]: Copyright needed

File: joint_trajectory_controller/src/joint_trajectory_controller_parameters.yaml [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    validation: {
+      lower_bounds: [0.1]
+    }
+  }
+  interpolation_method: {
+    type: string,
+    default_value: "splines",
+    description: "The type of interopoation to use if at all",
+    validation: {
+      one_of<>: [["splines", "none"]],
```
Comment (AndyZe) [inline joint_trajectory_controller/src/joint_trajectory_controller_parameters.yaml:71]: This is nice

File: joint_trajectory_controller/src/joint_trajectory_controller_parameters.yaml [hunk 4]
```diff
@@ -0,0 +1,73 @@
+joint_trajectory_controller:
+  joints: {
+    type: string_array,
+    default_value: [],
+    description: "Names of joints used by the controller",
```
Comment (AndyZe) [inline joint_trajectory_controller/src/joint_trajectory_controller_parameters.yaml:5]: So is this going to replace the list of interfaces in the URDF? I'm not sure about that.

It's nice having it in the URDF so it applies to all controllers, not just the JTC. It looks like what you have here will need to be duplicated a lot.

Comment (tylerjw) [reply joint_trajectory_controller/src/joint_trajectory_controller_parameters.yaml:5]: No, this just replaces the YAML parameter declaration and getting. This does not change anything about how interfaces from the URDF work.

File: joint_trajectory_controller/src/joint_trajectory_controller_parameters.yaml (hunk 3 above)
Comment (tylerjw) [reply joint_trajectory_controller/src/joint_trajectory_controller_parameters.yaml:71]: With this validation, if you accidentially set this value to something else you'll get a helpful error message that says something like this:

```Parameter failed validation: 'bla' set for 'interpolation_method' is not in the set ['splines', 'none']```

Comment (AndyZe) [reply joint_trajectory_controller/src/joint_trajectory_controller_parameters.yaml:71]: OK. Well, that case is already covered and it even allows for a default.

```
[[nodiscard]] inline InterpolationMethod from_string(const std::string & interpolation_method)
{
  if (interpolation_method.compare(InterpolationMethodMap.at(InterpolationMethod::NONE)) == 0)
  {
    return InterpolationMethod::NONE;
  }
  else if (
    !interpolation_method.compare(
      InterpolationMethodMap.at(InterpolationMethod::VARIABLE_DEGREE_SPLINE)) == 0)
  {
    return InterpolationMethod::VARIABLE_DEGREE_SPLINE;
  }
  // Default
  else
  {
    RCLCPP_INFO_STREAM(
      LOGGER,
      "No interpolation method parameter was given. Using the default, VARIABLE_DEGREE_SPLINE.");
    return InterpolationMethod::VARIABLE_DEGREE_SPLINE;
  }
}
```

File: joint_trajectory_controller/src/joint_trajectory_controller_parameters.yaml [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    default_value: 20.0,
+    description: "Rate status changes will be monitored",
+    validation: {
+      lower_bounds: [0.1]
+    }
+  }
+  interpolation_method: {
+    type: string,
+    default_value: "splines",
+    description: "The type of interopoation to use if at all",
```
Comment (AndyZe) [inline joint_trajectory_controller/src/joint_trajectory_controller_parameters.yaml:69]:
```suggestion
    description: "The type of interpolation to use, if any",
```

**Review by AndyZe — COMMENTED** [review]
Overall I think this looks like an improvement. There are more changes that will need to be made, though, like making `dof_` constant in `trajectory.cpp`.

File: joint_trajectory_controller/include/joint_trajectory_controller/validate_joint_trajectory_controller_parameters.hpp (hunk 2 above)
Comment (tylerjw) [reply joint_trajectory_controller/include/joint_trajectory_controller/validate_joint_trajectory_controller_parameters.hpp]: Done

File: joint_trajectory_controller/src/joint_trajectory_controller_parameters.yaml (hunk 3 above)
Comment (tylerjw) [reply joint_trajectory_controller/src/joint_trajectory_controller_parameters.yaml:63]: I think I should be able to remove that code from the `from_string` as with this you won't ever be able to get to the code path where it is invalid.

**Comment by tylerjw** [comment]
> Overall I think this looks like an improvement. There are more changes that will need to be made, though, like making `dof_` constant in `trajectory.cpp`.

Can you help me understand what you mean by this by pointing to the code you want me to update?

File: joint_trajectory_controller/package.xml (hunk 1 above)
Comment (tylerjw) [reply joint_trajectory_controller/package.xml]: Done

**Comment by tylerjw** [comment]
> I think we should talk about in the ros2_control WG meeting. Basically `dof_` should become constant (I think). That's going to change a bunch of other stuff in good ways. Then this PR can be applied on top of it.

https://github.com/ros-controls/ros2_controllers/issues/394

**Comment by tylerjw** [comment]
This passes all CI jobs but the ABI compatibility checks one which doesn't use a repos file so I wasn't able to add it as a source dependency. We'll bloom the package but it won't be ready for use without a repos file until the next sync.

File: joint_trajectory_controller/src/joint_trajectory_controller_parameters.yaml [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+    type: string_array,
+    default_value: [],
+    description: "Names of joints used by the controller",
+  }
+  command_interfaces: {
+    type: string_array,
+    default_value: [],
+    description: "Names of command interfaces to claim",
+    validation: {
+      unique<>: null,
```
Comment (destogl) [inline joint_trajectory_controller/src/joint_trajectory_controller_parameters.yaml:12]:
```suggestion
      unique<>: null,
      not_empty<>: null,
```

File: joint_trajectory_controller/src/joint_trajectory_controller_parameters.yaml [hunk 7]
```diff
@@ -0,0 +1,109 @@
+joint_trajectory_controller:
+  joints: {
+    type: string_array,
+    default_value: [],
+    description: "Names of joints used by the controller",
```
Comment (destogl) [inline joint_trajectory_controller/src/joint_trajectory_controller_parameters.yaml:5]:
```suggestion
    description: "Names of joints used by the controller",
    validation: {
      unique<>: null,
      not_empty<>: null,
    } 
```

File: joint_trajectory_controller/src/joint_trajectory_controller_parameters.yaml [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+    validation: {
+      unique<>: null,
+      not_empty<>: null,
```
Comment (destogl) [inline joint_trajectory_controller/src/joint_trajectory_controller_parameters.yaml:8]:
```suggestion
```

File: joint_trajectory_controller/src/joint_trajectory_controller_parameters.yaml [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+    } 
+  }
+  command_interfaces: {
+    type: string_array,
+    default_value: [],
+    description: "Names of command interfaces to claim",
+    validation: {
+      unique<>: null,
+      not_empty<>: null,
```
Comment (destogl) [inline joint_trajectory_controller/src/joint_trajectory_controller_parameters.yaml:17]:
```suggestion
```

File: joint_trajectory_controller/CMakeLists.txt [hunk 10]
```diff
@@ -29,12 +29,24 @@ foreach(Dependency IN ITEMS ${THIS_PACKAGE_INCLUDE_DEPENDS})
...
   src/trajectory.cpp
 )
 target_include_directories(${PROJECT_NAME} PRIVATE include)
-ament_target_dependencies(${PROJECT_NAME} ${THIS_PACKAGE_INCLUDE_DEPENDS})
+ament_target_dependencies(${PROJECT_NAME}
+  ${THIS_PACKAGE_INCLUDE_DEPENDS}
+)
+target_link_libraries(${PROJECT_NAME}
+  joint_trajectory_controller_parameters
+)
```
Comment (destogl) [inline joint_trajectory_controller/CMakeLists.txt:49]:
```suggestion
target_link_libraries(${PROJECT_NAME} joint_trajectory_controller_parameters)
ament_target_dependencies(${PROJECT_NAME} ${THIS_PACKAGE_INCLUDE_DEPENDS})
```

File: joint_trajectory_controller/src/joint_trajectory_controller.cpp [hunk 11]
```diff
@@ -46,32 +46,20 @@
...
-    std::string interpolation_string = auto_declare<std::string>(
-      "interpolation_method", interpolation_methods::InterpolationMethodMap.at(
-                                interpolation_methods::DEFAULT_INTERPOLATION));
-    interpolation_method_ = interpolation_methods::from_string(interpolation_string);
+    // Create the parameter listener and get the parameters
+    param_listener_ = std::make_shared<ParamListener>(get_node());
+    params_ = param_listener_->get_params();
+
+    // Set interpolation method from string parameter
+    interpolation_method_ = interpolation_methods::from_string(params_.interpolation_method);
```
[the remark this answers is not in the document]
Comment (destogl) [reply joint_trajectory_controller/src/joint_trajectory_controller.cpp:62]: Should be moved to the configure.

File: joint_trajectory_controller/src/joint_trajectory_controller.cpp [hunk 12]
```diff
@@ -98,10 +84,10 @@ JointTrajectoryController::command_interface_configuration() const
       dof_);
     std::exit(EXIT_FAILURE);
   }
-  conf.names.reserve(dof_ * command_interface_types_.size());
-  for (const auto & joint_name : command_joint_names_)
+  conf.names.reserve(dof_ * params_.command_interfaces.size());
+  for (const auto & joint_name : params_.joints)
```
Comment (destogl) [inline joint_trajectory_controller/src/joint_trajectory_controller.cpp:88]:
```suggestion
  for (const auto & joint_name : command_joint_name_)
```

**Review by bmagyar — APPROVED** [review]

**Pull request merged**